internal/jobservice/repository/job_service_in_memory.go

```diff
@@ -0,0 +1,113 @@
+package repository
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/G-Research/armada/internal/jobservice/configuration"
+	js "github.com/G-Research/armada/pkg/api/jobservice"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type JobStatus struct {
+	jobMap        map[string]*JobTable
+	jobLock       sync.RWMutex
+	subscribeMap  map[string]*string
+	subscribeLock sync.RWMutex
+}
+
+func NewJobStatus(jobMap map[string]*JobTable, subscribeMap map[string]*string) *JobStatus {
+	return &JobStatus{jobMap: jobMap, subscribeMap: subscribeMap}
+}
+
+type InMemoryJobServiceRepository struct {
+	jobStatus        *JobStatus
+	jobServiceConfig *configuration.JobServiceConfiguration
+}
+
+func NewInMemoryJobServiceRepository(jobMap *JobStatus, config *configuration.JobServiceConfiguration) *InMemoryJobServiceRepository {
+	return &InMemoryJobServiceRepository{jobStatus: jobMap, jobServiceConfig: config}
+}
+
+func (inMem *InMemoryJobServiceRepository) GetJobStatus(jobId string) (*js.JobServiceResponse, error) {
+	inMem.jobStatus.jobLock.RLock()
+	jobResponse, ok := inMem.jobStatus.jobMap[jobId]
+	inMem.jobStatus.jobLock.RUnlock()
+	if !ok {
+		return &js.JobServiceResponse{State: js.JobServiceResponse_JOB_ID_NOT_FOUND}, nil
+	}
+
+	return &jobResponse.jobResponse, nil
+}
+func (inMem *InMemoryJobServiceRepository) UpdateJobServiceDb(jobId string, jobTable *JobTable) error {
+	log.Infof("Updating JobId %s with State %s", jobId, jobTable.jobResponse.State)
+	inMem.jobStatus.jobLock.Lock()
+	defer inMem.jobStatus.jobLock.Unlock()
+	jobResponse, ok := inMem.jobStatus.jobMap[jobId]
+	if ok && jobResponse.jobResponse.State == jobTable.jobResponse.State {
+	} else {
+
+		inMem.jobStatus.jobMap[jobId] = jobTable
+	}
+	return nil
+}
+func (inMem *InMemoryJobServiceRepository) HealthCheck() bool {
+	return true
+}
+
+func (inMem *InMemoryJobServiceRepository) IsJobSetAlreadySubscribed(jobSetId string) bool {
+	inMem.jobStatus.subscribeLock.Lock()
+	defer inMem.jobStatus.subscribeLock.Unlock()
+	_, ok := inMem.jobStatus.subscribeMap[jobSetId]
+	if ok {
+		return true
+	}
+	inMem.jobStatus.subscribeMap[jobSetId] = &jobSetId
```

Comment:
I think this could be problematic because:
- we're marking that we are subscribed before we actually subscribe.  E.g. calling this method will cause us to become subscribed even though no subscription has acually ben made
- I wouldn't expect a function called `IsJobSetAlreadySubscribed` to mutate state.